Jasmine/Tetris/ViewModels/TetrisGameViewModel.swift

```diff
@@ -1,4 +1,165 @@
+                    line += getTileText(at: Coordinate(row: index, col: col))
+                }
+            } else {
+                for row in 0..<Constants.Tetris.rows {
+                    line += getTileText(at: Coordinate(row: row, col: index))
+                }
+            }
+
+            guard let validPhraseRange = getValidPhraseRange(line) else {
+                continue
+            }
+
+            for i in validPhraseRange {
+                let coordinate = searchByRow ? Coordinate(row: index, col: i) : Coordinate(row: i, col: index)
+                matchedCoordinates.insert(coordinate)
+            }
+            return matchedCoordinates
+        }
+        return nil
+    }
+
+    /// Gets the tile text at coordinate specified by `row` and `col`
+    /// returns " " if no tile is present so that phrases separated by gaps don't get matched
+    private func getTileText(at coordinate: Coordinate) -> String {
+        return tetrisGrid.getTileText(at: coordinate) ?? " "
+    }
+
+    /// Shifts all the tiles above `coordinates` 1 row down.
+    /// Starts from the row right above the coordinates so that it can break once an empty tile is encountered
+    fileprivate func shiftDownTiles(_ coordinates: Set<Coordinate>) {
+        var coordinatesToShift: [(from: Coordinate, to: Coordinate)] = []
+        for coordinate in coordinates {
+            for row in (0..<coordinate.row).reversed() {
+                let currentCoordinate = Coordinate(row: row, col: coordinate.col)
+                guard let text = tetrisGrid.removeTile(at: currentCoordinate) else {
+                    break
+                }
+                let newCoordinate = currentCoordinate.nextRow
+                tetrisGrid.addTile(at: newCoordinate, tileText: text)
+                coordinatesToShift.append((from: currentCoordinate, to: newCoordinate))
+            }
+        }
+        delegate?.animate(shiftTiles: coordinatesToShift)
+    }
+
+    // TODO: fetch from database to match valid phrase
+    private func getValidPhraseRange(_ line: String) -> CountableRange<Int>? {
+        let phrases = ["先发制人"]
+        for phrase in phrases {
+            let phraseLen = phrase.characters.count
+            for i in 0...line.characters.count - phraseLen {
+                let startIndex = line.index(line.startIndex, offsetBy: i)
+                if line[startIndex..<line.index(startIndex, offsetBy: phraseLen)] == phrase {
+                    return i..<i + phraseLen
+                }
+            }
+        }
+        return nil
+    }
+
+    // TODO: generate from database, base on existing grid
+    fileprivate func getRandomWord() -> String {
+        let words = "先发制人"
+        return String(words[words.index(words.startIndex,
+                                        offsetBy: Random.integer(toInclusive: UInt(words.characters.count)))])
+    }
+}
+
+extension TetrisGameViewModel: TetrisGameViewModelProtocol {
```

Comment:
Actually should bring some of the docs in the protocol here too. Pretty trouble to jump between code to find out what it does